book/invalidation.md

```diff
@@ -2254,10 +2296,26 @@ exercise.] is changed, it won't invalidate any layout fields (because
 these properties don't affect any layout fields) and so animations will
 once again skip layout entirely.
 
+
 ::: {.further}
-TODO
+
+CSS style depend on which elements a selector matches, and as the page
+changes, that may also need to be invalidated.[^our-browser] Browsers
+have clever algorithms to avoid redoing selector matching for every
+selector on the page. For example, Chromium constructs [*invalidation
+sets*][invalidation-set] for each selector, which tell it which
+elements to recheck each selector on. New selectors such as `:has()`
+make invalidation [more complicated][has-invalidation], but this
+complexity is necessary for fast re-styles.
+
 :::
 
+[invalidation-set]: https://chromium.googlesource.com/chromium/src/+/HEAD/third_party/blink/renderer/core/css/style-invalidation.md?pli=1#
+[has-invalidation]: https://blogs.igalia.com/blee/posts/2023/05/31/how-blink-invalidates-styles-when-has-in-use.html
+[^our-browser]: The reason this isn't necessary in our browser is just
+    that our browser supports too few CSS selectors and too few DOM
+    APIs for any selector matches to actually change!
```

Comment:
There is `:focus` and innerHTML, which do cause this. I'd say instead that our browser has too few features for such advanced techniques to make sense.